HW 1!
Emma Brown and I worked together on this assignment, and it was super fun. We used chatgpt once to help with our code, I will note that later on. 

In-class work from tabular_data_02

In [1]:
import pandas as pd
import numpy as np

In [2]:
import os
os.chdir('/Users/katielark/Downloads/data_viz')
df = pd.read_csv('dhs_service_records_synthesized_final.csv')

Which recipient has the highest number of records among all recipients?
Answer: 178128

In [3]:
df.groupby("MCI_UNIQ_ID").count().sort_values(by="synthetic_data", ascending=False)

synthetic_data  CALDR_YR  DATE_OF_EVENT  service  GEO_AREA  age  \
MCI_UNIQ_ID                                                                    
178128                   89        89             89       89         0   89   
392121                   89        89             89       89         0   89   
306750                   88        88             88       88         0   88   
221385                   85        85             85       85         0   85   
121055                   85        85             85       85         0   85   
...                     ...       ...            ...      ...       ...  ...   
82846                     1         1              1        1         0    1   
302861                    1         1              1        1         0    1   
143362                    1         1              1        1         0    1   
112358                    1         1              1        1         0    1   
109229                    1         1              1        1         0    1   

             DOB  DOD  GENDER  GENDER_IDENTITY  SEX_ORIENT  LEGAL_SEX  RACE  \
MCI_UNIQ_ID                                                                   
178128         0    0      89                0           0          0    89   
392121         0    0      89                0           0          0    89   
306750         0    0      88                0           0          0    88   
221385         0    0      85                0           0          0    85   
121055         0    0      85                0           0          0    85   
...          ...  ...     ...              ...         ...        ...   ...   
82846          0    0       1                0           0          0     1   
302861         0    0       1                0           0          0     1   
143362         0    0       1                0           0          0     1   
112358         0    0       1                0           0          0     1   
109229         0    0       1                0           0          0     1   

             ETHNICITY  LIVING_ARRANGEMENT  EMPLOYMENT_STATUS  MARITAL_STATUS  \
MCI_UNIQ_ID                                                                     
178128              89                   0                  0              89   
392121              89                   0                  0              89   
306750              88                   0                  0              88   
221385              85                   0                  0              85   
121055              85                   0                  0              85   
...                ...                 ...                ...             ...   
82846                1                   0                  0               1   
302861               1                   0                  0               1   
143362               1                   0                  0               1   
112358               1                   0                  0               1   
109229               1                   0                  0               1   

             EDUCATION_LEVEL  VETERAN_FLAG  
MCI_UNIQ_ID                                 
178128                    89             0  
392121                    89             0  
306750                    88             0  
221385                    85             0  
121055                    85             0  
...                      ...           ...  
82846                      1             0  
302861                     1             0  
143362                     1             0  
112358                     1             0  
109229                     1             0  

[533799 rows x 19 columns]

How many recipients used the services in January?
Answer: 457593

In [4]:
df[df["DATE_OF_EVENT"] == "01/31/2021"].nunique()

synthetic_data             1
MCI_UNIQ_ID           457593
CALDR_YR                   1
DATE_OF_EVENT              1
service                   22
GEO_AREA                   0
age                      112
DOB                        0
DOD                        0
GENDER                     3
GENDER_IDENTITY            0
SEX_ORIENT                 0
LEGAL_SEX                  0
RACE                       7
ETHNICITY                  3
LIVING_ARRANGEMENT         0
EMPLOYMENT_STATUS          0
MARITAL_STATUS             7
EDUCATION_LEVEL            8
VETERAN_FLAG               0
dtype: int64

For each month, how many recipients used services?
Answer: (see dataframe below)

In [5]:
df.groupby("DATE_OF_EVENT").nunique()

synthetic_data  MCI_UNIQ_ID  CALDR_YR  service  GEO_AREA  age  \
DATE_OF_EVENT                                                                  
01/31/2021                  1       457593         1       22         0  112   
02/28/2021                  1       461207         1       22         0  112   
03/31/2021                  1       465763         1       22         0  112   
04/30/2021                  1       469090         1       22         0  112   
05/31/2021                  1       467082         1       22         0  112   
06/30/2021                  1       470430         1       22         0  112   
07/31/2021                  1       473887         1       22         0  112   
08/31/2021                  1       477523         1       22         0  112   
09/30/2021                  1       484128         1       22         0  112   
10/31/2021                  1       489308         1       22         0  112   
11/30/2021                  1       494557         1       22         0  112   
12/31/2021                  1       495124         1       22         0  112   

               DOB  DOD  GENDER  GENDER_IDENTITY  SEX_ORIENT  LEGAL_SEX  RACE  \
DATE_OF_EVENT                                                                   
01/31/2021       0    0       3                0           0          0     7   
02/28/2021       0    0       3                0           0          0     7   
03/31/2021       0    0       3                0           0          0     7   
04/30/2021       0    0       3                0           0          0     7   
05/31/2021       0    0       3                0           0          0     7   
06/30/2021       0    0       3                0           0          0     7   
07/31/2021       0    0       3                0           0          0     7   
08/31/2021       0    0       3                0           0          0     7   
09/30/2021       0    0       3                0           0          0     7   
10/31/2021       0    0       3                0           0          0     7   
11/30/2021       0    0       3                0           0          0     7   
12/31/2021       0    0       3                0           0          0     7   

               ETHNICITY  LIVING_ARRANGEMENT  EMPLOYMENT_STATUS  \
DATE_OF_EVENT                                                     
01/31/2021             3                   0                  0   
02/28/2021             3                   0                  0   
03/31/2021             3                   0                  0   
04/30/2021             3                   0                  0   
05/31/2021             3                   0                  0   
06/30/2021             3                   0                  0   
07/31/2021             3                   0                  0   
08/31/2021             3                   0                  0   
09/30/2021             3                   0                  0   
10/31/2021             3                   0                  0   
11/30/2021             3                   0                  0   
12/31/2021             3                   0                  0   

               MARITAL_STATUS  EDUCATION_LEVEL  VETERAN_FLAG  
DATE_OF_EVENT                                                 
01/31/2021                  7                8             0  
02/28/2021                  7                8             0  
03/31/2021                  7                8             0  
04/30/2021                  7                8             0  
05/31/2021                  7                8             0  
06/30/2021                  7                8             0  
07/31/2021                  7                8             0  
08/31/2021                  7                8             0  
09/30/2021                  7                8             0  
10/31/2021                  7                8             0  
11/30/2021                  7                8             0  
1

Given an ID, how many months did the particular recipient (by ID) use the service? e.g., 449747
Answer: (see dataframe below)

In [6]:
df[df["MCI_UNIQ_ID"] == 449747].nunique()

synthetic_data         1
MCI_UNIQ_ID            1
CALDR_YR               1
DATE_OF_EVENT         10
service                1
GEO_AREA               0
age                    1
DOB                    0
DOD                    0
GENDER                 1
GENDER_IDENTITY        0
SEX_ORIENT             0
LEGAL_SEX              0
RACE                   1
ETHNICITY              1
LIVING_ARRANGEMENT     0
EMPLOYMENT_STATUS      0
MARITAL_STATUS         1
EDUCATION_LEVEL        1
VETERAN_FLAG           0
dtype: int64

Do the statistics on the columns make sense? e.g., are there any inconsistencies between age and education level?
Answer: There are 112 unique ages, which is inconsistent with average human life span. The dataset contains a high number of younger children. This is consistent with the education data demonstrating a high portion of recipients only completing elementary school. 

In-class work from tabular_data_03

In [7]:
df.rename(columns=str.lower, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7116134 entries, 0 to 7116133
Data columns (total 20 columns):
 #   Column              Dtype  
---  ------              -----  
 0   synthetic_data      object 
 1   mci_uniq_id         int64  
 2   caldr_yr            int64  
 3   date_of_event       object 
 4   service             object 
 5   geo_area            float64
 6   age                 int64  
 7   dob                 float64
 8   dod                 float64
 9   gender              object 
 10  gender_identity     float64
 11  sex_orient          float64
 12  legal_sex           float64
 13  race                object 
 14  ethnicity           object 
 15  living_arrangement  float64
 16  employment_status   float64
 17  marital_status      object 
 18  education_level     object 
 19  veteran_flag        float64
dtypes: float64(9), int64(3), object(8)
memory usage: 1.1+ GB


In [8]:
df.rename(columns = {'mci_uniq_id': 'id', 
                    'date_of_event': 'date', 
                    'marital_status': 'marital', 
                    'education_level':'education'}, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7116134 entries, 0 to 7116133
Data columns (total 20 columns):
 #   Column              Dtype  
---  ------              -----  
 0   synthetic_data      object 
 1   id                  int64  
 2   caldr_yr            int64  
 3   date                object 
 4   service             object 
 5   geo_area            float64
 6   age                 int64  
 7   dob                 float64
 8   dod                 float64
 9   gender              object 
 10  gender_identity     float64
 11  sex_orient          float64
 12  legal_sex           float64
 13  race                object 
 14  ethnicity           object 
 15  living_arrangement  float64
 16  employment_status   float64
 17  marital             object 
 18  education           object 
 19  veteran_flag        float64
dtypes: float64(9), int64(3), object(8)
memory usage: 1.1+ GB


Identify the top services by number of recipients. You can choose to do by other measures (e.g., total usage months). 
Answer: Income support, DHS services, Mental health services, housing services

In [9]:
df.groupby("service").count().sort_values(by= "id", ascending = False)

synthetic_data       id  \
service                                                                       
Individuals_Receiving_Income_Supports                      5450993  5450993   
Individuals_Receiving_DHS_Services                          626722   626722   
Individuals_Receiving_Mental_Health_Services                324241   324241   
Individuals_Receiving_Homelessness_and_Housing_...          148968   148968   
Older_Adults_Receiving_Services                             131153   131153   
Individuals_Receiving_Substance_Use_Disorder_Se...           74737    74737   
Individuals_Receiving_Intellectual_Disability_S...           74145    74145   
Parents_Receiving_Child_Welfare_Services                     54035    54035   
Children_Receiving_Child_Welfare_Services                    50373    50373   
Children_Receiving_Early_Intervention_Services               45923    45923   
Individuals_Receiving_Family_Strengthening_Prog...           28110    28110   
Jail_Population                                              26987    26987   
Families_Receiving_Child_Welfare_Services                    23473    23473   
Children_in_Care                                             18442    18442   
Homeless_Population                                          13950    13950   
Mental_Health_Crises                                          7245     7245   
Involuntary_Commitments                                       7061     7061   
Children_Receiving_DHS_Funded_Out_of_School_Pro...            4606     4606   
Children_Attending_Early_Childhood_Programs_Man...            2302     2302   
Overdoses                                                     1319     1319   
Suicides                                                      1177     1177   
Homicides                                                      172      172   

                                                    caldr_yr     date  \
service                                                                 
Individuals_Receiving_Income_Supports                5450993  5450993   
Individuals_Receiving_DHS_Services                    626722   626722   
Individuals_Receiving_Mental_Health_Services          324241   324241   
Individuals_Receiving_Homelessness_and_Housing_...    148968   148968   
Older_Adults_Receiving_Services                       131153   131153   
Individuals_Receiving_Substance_Use_Disorder_Se...     74737    74737   
Individuals_Receiving_Intellectual_Disability_S...     74145    74145   
Parents_Receiving_Child_Welfare_Services               54035    54035   
Children_Receiving_Child_Welfare_Services              50373    50373   
Children_Receiving_Early_Intervention_Services         45923    45923   
Individuals_Receiving_Family_Strengthening_Prog...     28110    28110   
Jail_Population                                        26987    26987   
Families_Receiving_Child_Welfare_Services              23473    23473   
Children_in_Care                                       18442    18442   
Homeless_Population                                    13950    13950   
Mental_Health_Crises                                    7245     7245   
Involuntary_Commitments                                 7061     7061   
Children_Receiving_DHS_Funded_Out_of_School_Pro...      4606     4606   
Children_Attending_Early_Childhood_Programs_Man...      2302     2302   
Overdoses                                               1319     1319   
Suicides                                                1177     1177   
Homicides                                                172      172   

                                                    geo_area      age  dob  \
service                                                                      
Individuals_Receiving_Income_Supports                      0  5450993    0   
Individuals_Receiving_DHS_Services                         0   626722    0   
Individuals_Receiving_Mental_Health_Services               0   324241    0   
Individuals_

In [11]:
df = df.astype({'id':'category', 
                'service':'category',
                'age': 'int',
                'gender':'category',
                'race':'category',
                'marital':'category',
                'education':'category'})

Define a python function that computes characteristic attributes for services.
Input: the service recipient event dataset.
Output: a data frame, with rows for services and columns for various attributes of the services.
Design your own attributes, see examples of attributes for recipients.

In [12]:
def get_services_attribute(df):
    services = df.groupby(df['service'].cat.remove_unused_categories()).agg(  ### remove unused categories 
        num_recipients = ('service', 'count'),
        first_date = ('date', 'min'), 
        last_date = ('date', 'max'),
        avg_age = ('age', 'mean'),
    ).reset_index()

    return services

Create a new dataframe that hold attributes for services, row keyed by service
and columns contain attributes that output by the function you created.

In [13]:
services_new = get_services_attribute(df)

/var/folders/00/yzv7lrx12xq5_g14dw7m_2180000gn/T/ipykernel_22692/4099685790.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  services = df.groupby(df['service'].cat.remove_unused_categories()).agg(  ### remove unused categories


In [14]:
services_new

service  num_recipients  \
0   Children_Attending_Early_Childhood_Programs_Ma...            2302   
1           Children_Receiving_Child_Welfare_Services           50373   
2   Children_Receiving_DHS_Funded_Out_of_School_Pr...            4606   
3      Children_Receiving_Early_Intervention_Services           45923   
4                                    Children_in_Care           18442   
5           Families_Receiving_Child_Welfare_Services           23473   
6                                 Homeless_Population           13950   
7                                           Homicides             172   
8                  Individuals_Receiving_DHS_Services          626722   
9   Individuals_Receiving_Family_Strengthening_Pro...           28110   
10  Individuals_Receiving_Homelessness_and_Housing...          148968   
11              Individuals_Receiving_Income_Supports         5450993   
12  Individuals_Receiving_Intellectual_Disability_...           74145   
13       Individuals_Receiving_Mental_Health_Services          324241   
14  Individuals_Receiving_Substance_Use_Disorder_S...           74737   
15                            Involuntary_Commitments            7061   
16                                    Jail_Population           26987   
17                               Mental_Health_Crises            7245   
18                    Older_Adults_Receiving_Services          131153   
19                                          Overdoses            1319   
20           Parents_Receiving_Child_Welfare_Services           54035   
21                                           Suicides            1177   

    first_date   last_date    avg_age  
0   01/31/2021  12/31/2021   4.297133  
1   01/31/2021  12/31/2021   9.746273  
2   01/31/2021  12/31/2021   9.536691  
3   01/31/2021  12/31/2021   1.592971  
4   01/31/2021  12/31/2021   8.311300  
5   01/31/2021  12/31/2021  41.063520  
6   01/31/2021  12/31/2021  35.731756  
7   01/31/2021  12/31/2021  29.697674  
8   01/31/2021  12/31/2021  36.936773  
9   01/31/2021  12/31/2021  25.745535  
10  01/31/2021  12/31/2021  37.300239  
11  01/31/2021  12/31/2021  31.094901  
12  01/31/2021  12/31/2021  35.983114  
13  01/31/2021  12/31/2021  34.380174  
14  01/31/2021  12/31/2021  42.891071  
15  01/31/2021  12/31/2021  41.709956  
16  01/31/2021  12/31/2021  40.132064  
17  01/31/2021  12/31/2021  37.834783  
18  01/31/2021  12/31/2021  79.297126  
19  01/31/2021  12/31/2021  42.628506  
20  01/31/2021  12/31/2021  40.832775  
21  01/31/2021  12/31/2021  27.717077

Experiment aggregation with or without reset_index() and see what happens.
Answer: The indices on the left are removed and the column "services" is not in line with the other columns. 

In [15]:
def get_services_attribute_wri(df):
    services = df.groupby(df['service'].cat.remove_unused_categories()).agg(  ### remove unused categories 
        num_recipients = ('service', 'count'),
        first_date = ('date', 'min'), 
        last_date = ('date', 'max'),
        avg_age = ('age', 'mean'),
    )

    return services

In [16]:
services_new_wri = get_services_attribute_wri(df)

/var/folders/00/yzv7lrx12xq5_g14dw7m_2180000gn/T/ipykernel_22692/3766694709.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  services = df.groupby(df['service'].cat.remove_unused_categories()).agg(  ### remove unused categories


In [17]:
services_new_wri

num_recipients  \
service                                                              
Children_Attending_Early_Childhood_Programs_Man...            2302   
Children_Receiving_Child_Welfare_Services                    50373   
Children_Receiving_DHS_Funded_Out_of_School_Pro...            4606   
Children_Receiving_Early_Intervention_Services               45923   
Children_in_Care                                             18442   
Families_Receiving_Child_Welfare_Services                    23473   
Homeless_Population                                          13950   
Homicides                                                      172   
Individuals_Receiving_DHS_Services                          626722   
Individuals_Receiving_Family_Strengthening_Prog...           28110   
Individuals_Receiving_Homelessness_and_Housing_...          148968   
Individuals_Receiving_Income_Supports                      5450993   
Individuals_Receiving_Intellectual_Disability_S...           74145   
Individuals_Receiving_Mental_Health_Services                324241   
Individuals_Receiving_Substance_Use_Disorder_Se...           74737   
Involuntary_Commitments                                       7061   
Jail_Population                                              26987   
Mental_Health_Crises                                          7245   
Older_Adults_Receiving_Services                             131153   
Overdoses                                                     1319   
Parents_Receiving_Child_Welfare_Services                     54035   
Suicides                                                      1177   

                                                    first_date   last_date  \
service                                                                      
Children_Attending_Early_Childhood_Programs_Man...  01/31/2021  12/31/2021   
Children_Receiving_Child_Welfare_Services           01/31/2021  12/31/2021   
Children_Receiving_DHS_Funded_Out_of_School_Pro...  01/31/2021  12/31/2021   
Children_Receiving_Early_Intervention_Services      01/31/2021  12/31/2021   
Children_in_Care                                    01/31/2021  12/31/2021   
Families_Receiving_Child_Welfare_Services           01/31/2021  12/31/2021   
Homeless_Population                                 01/31/2021  12/31/2021   
Homicides                                           01/31/2021  12/31/2021   
Individuals_Receiving_DHS_Services                  01/31/2021  12/31/2021   
Individuals_Receiving_Family_Strengthening_Prog...  01/31/2021  12/31/2021   
Individuals_Receiving_Homelessness_and_Housing_...  01/31/2021  12/31/2021   
Individuals_Receiving_Income_Supports               01/31/2021  12/31/2021   
Individuals_Receiving_Intellectual_Disability_S...  01/31/2021  12/31/2021   
Individuals_Receiving_Mental_Health_Services        01/31/2021  12/31/2021   
Individuals_Receiving_Substance_Use_Disorder_Se...  01/31/2021  12/31/2021   
Involuntary_Commitments                             01/31/2021  12/31/2021   
Jail_Population                                     01/31/2021  12/31/2021   
Mental_Health_Crises                                01/31/2021  12/31/2021   
Older_Adults_Receiving_Services                     01/31/2021  12/31/2021   
Overdoses                                           01/31/2021  12/31/2021   
Parents_Receiving_Child_Welfare_Services            01/31/2021  12/31/2021   
Suicides                                            01/31/2021  12/31/2021   

                                                      avg_age  
service                                                        
Children_Attending_Early_Childhood_Programs_Man...   4.297133  
Children_Receiving_Child_Welfare_Services            9.746273  
Children_Receiving_DHS_Funded_Out_of_School_Pro...   9.536691  
Children_Receiving_Early_Intervention_Services       1.592971  
Children_in_Care                                     8.311300  
Families_Receiving_Child_Welfare_Services         

Homework!

Modify the function “attributes of service” from your in-class work into a new function that take the service usage data frame as input, and return a dataframe, with rows for services, and additional columns for total number of recipients and average monthly number of recipients.
Note: We used chatgpt on this code. We originally calculated the avg_monthly inside the agg() command. Chatgpt suggested we move the column to outside the command and add the df[]. This makes sense because the columns total_recipients and distinct_month had not been defined. 
Notably, the avg_monthly uses the total_recipients and not the total_unique_recipients column (which have very different values)

In [18]:
def get_services_attribute(df):
    services = df.groupby(df['service'].cat.remove_unused_categories()).agg(
        total_recipients = ('service', 'count'),
        total_unique_recpients = ('id', 'nunique'),
        distinct_month = ('month', 'nunique'),
    ).reset_index()
    
    services['avg_monthly'] = services['total_recipients'] / services['distinct_month']
    
    return services

Call your function to generate the report on services, display the results. 

In [20]:
df["month_str"] = df["date"].str[:2] + df["date"].str[-5:]

In [21]:
df['date'] = pd.to_datetime(df['date'])
df['month'] = df["date"].dt.month_name()

In [22]:
services_exp = get_services_attribute(df)

/var/folders/00/yzv7lrx12xq5_g14dw7m_2180000gn/T/ipykernel_22692/2940856386.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  services = df.groupby(df['service'].cat.remove_unused_categories()).agg(


In [23]:
services_exp

service  total_recipients  \
0   Children_Attending_Early_Childhood_Programs_Ma...              2302   
1           Children_Receiving_Child_Welfare_Services             50373   
2   Children_Receiving_DHS_Funded_Out_of_School_Pr...              4606   
3      Children_Receiving_Early_Intervention_Services             45923   
4                                    Children_in_Care             18442   
5           Families_Receiving_Child_Welfare_Services             23473   
6                                 Homeless_Population             13950   
7                                           Homicides               172   
8                  Individuals_Receiving_DHS_Services            626722   
9   Individuals_Receiving_Family_Strengthening_Pro...             28110   
10  Individuals_Receiving_Homelessness_and_Housing...            148968   
11              Individuals_Receiving_Income_Supports           5450993   
12  Individuals_Receiving_Intellectual_Disability_...             74145   
13       Individuals_Receiving_Mental_Health_Services            324241   
14  Individuals_Receiving_Substance_Use_Disorder_S...             74737   
15                            Involuntary_Commitments              7061   
16                                    Jail_Population             26987   
17                               Mental_Health_Crises              7245   
18                    Older_Adults_Receiving_Services            131153   
19                                          Overdoses              1319   
20           Parents_Receiving_Child_Welfare_Services             54035   
21                                           Suicides              1177   

    total_unique_recpients  distinct_month    avg_monthly  
0                      349              12     191.833333  
1                     7306              12    4197.750000  
2                     1198              12     383.833333  
3                     6752              12    3826.916667  
4                     2237              12    1536.833333  
5                     3118              12    1956.083333  
6                     3415              12    1162.500000  
7                      172              12      14.333333  
8                    82122              12   52226.833333  
9                     2986              12    2342.500000  
10                   21823              12   12414.000000  
11                  502530              12  454249.416667  
12                    6360              12    6178.750000  
13                   55263              12   27020.083333  
14                   12812              12    6228.083333  
15                    5010              12     588.416667  
16                    6922              12    2248.916667  
17                    7245              12     603.750000  
18                   16759              12   10929.416667  
19                    1319              12     109.916667  
20                    7078              12    4502.916667  
21                    1177              12      98.083333

Write another function that produces the numbers of recipients by month.

In [24]:
def get_recipient_month_attribute(df):
    service_month = df.groupby(['service', 'month']).agg(
        num_id = ('id', 'count'),
    ).reset_index()

    return service_month

What conclusions can you draw about the services in all, and about individual services?
Answer: The services are necessary and utilized. Services are utilized relatively evenly across months, with November having the most utilization and January having the least. Most recipients of services not catering directly to children are in their late 30s, early 40s. 